notification-forwarder/notification-forwarder.py

```diff
@@ -0,0 +1,464 @@
+	#
+	# while True:
+	#   print("Notifying...")
+	#   iface.Notify('foo', 0, '', 'summary', 'body', [], {}, 10000)
+	#   time.sleep(300)
+        proxy = self.private_bus.get_object("org.freedesktop.Notifications", "/org/freedesktop/Notifications")
+        return dbus.Interface(proxy, "org.freedesktop.Notifications")
+
+    def run(self):
+        GLib.MainLoop().run()
+
+    def schedule(self, worker, queue, content):
+        if worker.is_alive():
+            queue.put(content, timeout=5)
+        else:
+            self.log.error(f"{worker.__class__.__name__} dead. Exiting...")
+            sys.exit(2)
+
+    def schedule_qrexec(self, content):
+        return self.schedule(self.qrexec_worker, self.qrexec_queue, content)
+
+    @dbus.service.method("org.freedesktop.Notifications", in_signature="",
+                         out_signature="as")
+    def GetCapabilities(self):
+        #always handle locally
+        self.log.debug("GetCapabilities")
+        return self.capabilities
+
+    @dbus.service.method("org.freedesktop.Notifications",
+                         in_signature="susssasa{sv}i",
+                         out_signature="u")
+    def Notify(self, app_name, notification_id, app_icon,
+               summary, body, actions, hints, expire_timeout):
+        notification = Notification(
+            app_name=app_name,
+            notification_id=notification_id,
+            app_icon=app_icon,
+            summary=summary,
+            body=body,
+            actions=actions,
+            hints=hints,
+            expire_timeout=expire_timeout,
+            )
+        self.log.debug("Notify")
+
+        if self.should_forward(notification):
+            self.log.debug(f"Forwarding: {notification}")
+            server_id = self.forward(notification)
+        else:
+            self.log.debug(f"Handling locally: {notification}")
+
+            if notification_id == 0:
+                #new ID required
+                server_id = self.allocate_id()
+                local_id = 0
+            else:
+                #previously used ID
+                server_id = notification_id
+                local_id = self.server2local_id.get(server_id, 0)
+
+            local_id = self.iface.Notify(app_name, local_id, app_icon, summary,
+                                         body, actions, hints, expire_timeout,
+                                         signature="susssasa{sv}i")
+            self.update_id(server_id, local_id)
+
+        return server_id
+
+    def should_forward(self, notification):
+        """ Returns true, if the given notification is meant to be forwarded to the remote VM. """
+        if self.force_mode:
+            return True
+        if self.local_mode:
+            return False
+
+        # regular mode
+        # We cannot handle actions as they require a backchannel which might not be so wise having from
+        # a security point of view. So we handle those locally.
```

Comment:
If the notification qube is the same as the GUI qube (which I imagine is a useful default), then a backchannel is not a problem.